File: HW/HW10/hw10.py
from ast import List
import collections
from typing import Optional


class TreeNode(object):
     def __init__(self, val=0, left=None, right=None):
        #python definiiton for treenodes
         self.val = val
         self.left = left
         self.right = right

class hw10:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = [] #final array, the rsulting answer
        queue = collections.deque() #initializes a queue.
        queue.append(root)

        #run BFS while it isnt empty
        while queue: 
            queueLen = len(queue) #get # of nodes in queue currently, loop trhough every single values
            # helps make sure we go through one level at a time
            level = [] #eachlevel is a list, gets added in the answer
            for i in range(queueLen):
                #fifo; this is what we want
                node = queue.popleft() 
                # note : pop():- This function is used to delete an argument from the right end of the deque.
                # popleft():- This function is used to delete an argument from the left end of the deque.

                if node: #checks non null before adding
                    level.append(node.val)
                    queue.append(node.left)
                    queue.append(node.right)
            if level:
                ans.append(level)
            #loop iterates every node in a single level, after we added that level, we add it to the result, keep running outside while loop until there are no nodes left in queue
        return ans

File: HW/HW10/hw10.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class hw10 {
    public class TreeNode { //java definition for treenodes
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {}
   
        TreeNode(int val) { this.val = val; }
   
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
     }

     public List<List<Integer>> levelOrder(TreeNode root) {
       List<List<Integer>> ans = new ArrayList<>();
       if (root == null) {return ans;}

       Queue<TreeNode> queue = new LinkedList<>();
       queue.add(root);

       while(!queue.isEmpty()){
        //while loop that runs when the queue isn't empty, this pops and adds values to the queue until queue is empty , that is when we are done and return answer
            List<Integer> level = new ArrayList<>();
            int len = queue.size();
            for( int i = 0 ; i < len; i++){
                TreeNode node = queue.poll();
                //poll: Retrieves and removes the head of this queue, or returns null if this queue is empty.
                level.add(node.val);
                //adds the value to level arraylist
                if(node.left!= null){
                    queue.add(node.left);
                    //if the nodes are not null, then add the childs of the node to the queue
                }
                if(node.right!=null){
                    queue.add(node.right);
                }
            }
            ans.add(level);
            //adds level arraylist to answer list
       }
       return ans;
    }


}
